src/lib.rs

```diff
@@ -546,11 +549,39 @@ pub trait Helper
 where
     Self: Completer + Hinter + Highlighter + Validator,
 {
+    /// Update helper when line has been modified.
+    ///
+    /// This is the first-called function just after the editing is done,
+    /// before all other functions within [Completer], [Hinter], [Highlighter], and [Validator].
+    ///
+    /// You can put the tokenizer/parser here so that other APIs can directly use
+    /// results generate here, and reduce the overhead.
+    fn update_after_edit(&mut self, line: &str, pos: usize, forced_refresh: bool) {
```

Comment:
I would have prefer a dedicated component and an API compatible with tree-sitter incremental parsing.
See https://github.com/gwenn/rustyline-notes/blob/master/src/syntax.md

Reply:
And this may conflict with another feature: autocorrection.
See https://github.com/kkawakam/rustyline/issues/593#issuecomment-1014780747
And https://github.com/kkawakam/rustyline/issues/656#issuecomment-1299641724